internal/tools/orchestrator/dbclient/hpa.go

```diff
@@ -0,0 +1,165 @@
+	Reason        string      `json:"reason,omitempty"`
+	Message       string      `json:"message,omitempty"`
+}
+
+func (HPAEventInfo) TableName() string {
+	return "ps_v2_runtime_hpa_events"
+}
+
+func (db *DBClient) CreateRuntimeHPA(runtimeHPA *RuntimeHPA) error {
+	return db.Save(runtimeHPA).Error
+}
+
+func (db *DBClient) UpdateRuntimeHPA(runtimeHPA *RuntimeHPA) error {
+	if err := db.Model(&RuntimeHPA{}).Where("rule_id = ?", runtimeHPA.RuleID).Update(runtimeHPA).Error; err != nil {
+		return errors.Wrapf(err, "failed to update runtime hpa rule, id: %v", runtimeHPA.RuleID)
+	}
+	return nil
+}
+
+// if not found, return (nil, error)
+func (db *DBClient) GetRuntimeHPAByServices(id spec.RuntimeUniqueId, services []string) ([]RuntimeHPA, error) {
+	var runtimeHPAs []RuntimeHPA
+	if len(services) > 0 {
+		if err := db.
+			Where("application_id = ? AND workspace = ? AND runtime_name = ? AND service_name in (?)", id.ApplicationId, id.Workspace, id.Name, services).
+			Find(&runtimeHPAs).Error; err != nil {
+			return nil, errors.Wrapf(err, "failed to get runtime hpa rule for runtime %+v for services: %v", id, services)
+		}
+	} else {
+		if err := db.
+			Where("application_id = ? AND workspace = ? AND runtime_name = ? ", id.ApplicationId, id.Workspace, id.Name).
+			Find(&runtimeHPAs).Error; err != nil {
+			return nil, errors.Wrapf(err, "failed to get runtime hpa rule for runtime: %+v", id)
+		}
+	}
+	return runtimeHPAs, nil
+}
+
+func (db *DBClient) DeleteRuntimeHPAByRuleId(ruleId string) error {
+	if err := db.
+		Where("rule_id = ?", ruleId).
+		Delete(&RuntimeHPA{}).Error; err != nil {
+		return errors.Wrapf(err, "failed to delete runtime hpa rule for rule_id: %v", ruleId)
+	}
+	return nil
+}
+
+func (db *DBClient) GetRuntimeHPARuleByRuleId(ruleId string) (*RuntimeHPA, error) {
+	var runtimeHPA RuntimeHPA
+	if err := db.
+		Where("rule_id = ?", ruleId).
+		Find(&runtimeHPA).Error; err != nil {
+		return nil, errors.Wrapf(err, "failed to get runtime hpa rule for rule_id: %v", ruleId)
+	}
+	return &runtimeHPA, nil
+}
+
+func (db *DBClient) GetRuntimeHPARulesByRuntimeId(runtimeId uint64) ([]RuntimeHPA, error) {
+	var runtimeHPAs []RuntimeHPA
+	if err := db.
+		Where("runtime_id = ?", runtimeId).
+		Find(&runtimeHPAs).Error; err != nil {
+		return nil, errors.Wrapf(err, "failed to get runtime hpa rule for runtime_id: %v", runtimeId)
+	}
+	return runtimeHPAs, nil
+}
+
+func (db *DBClient) CreateHPAEventInfo(hpaEvent *HPAEventInfo) error {
+	return db.Save(hpaEvent).Error
+}
+
+// if not found, return (nil, error)
+func (db *DBClient) GetRuntimeHPAEventsByServices(runtimeId uint64, services []string) ([]HPAEventInfo, error) {
+	var hpaEvents []HPAEventInfo
+	if len(services) > 0 {
+		//  select * from ps_v2_runtime_hpa_events  where runtime_id = '143' AND service_name in ('go-demo','abc') order by updated_at desc limit 20;
+		if err := db.
+			Where("runtime_id = ? AND service_name in (?)", runtimeId, services).Order("updated_at desc").Limit(20).
```

Comment:
magic number 20

Reply:
For normal case, hpa event rarely, so get recently 20 events for per service. get recently 100 events for per runtimes.

Set this two as constant: 
```go
// ErdaHPARecentlyEventsMaxToListForRuntimeDefault define max hpa events for recently to list for runtime
ErdaHPARecentlyEventsMaxToListForRuntimeDefault int32 = 100
// ErdaHPARecentlyEventsMaxToListForServiceDefault define max hpa events for recently to list for service
ErdaHPARecentlyEventsMaxToListForServiceDefault int32 = 20
```